### ISSUE TITLE
CRAN issues

### ISSUE THREAD

**Issue by eddelbuettel** [eddelbuettel/nanotime#120]
We apparently have [two CRAN issues](https://cloud.r-project.org/web/checks/check_results_nanotime.html) outside the long-fixed 'do not specify C++11 unless needed' we addressed ages ago.

One is just a typo. R-devel checks more carefully for 'isolated' pairs of `{}` we have one `code{+}` that wants to be `\code{+}`.  I will fix that in a second.  (One aside is that we do _something_ that upsets `roxygen2` on my machine.  It errors with ` "S3 method {.arg {name}} needs @export or @exportS3method tag")` and dies.)

The other is that of late Brian Ripley is chasing more uses of functions from 'internal' headers and our use of `SET_S4_OBJECT` is flagged, twice.  The function is a simple one-liner so we could internalize, or we could replace the S4 creation with a call from C++ to R and call a proper `new(...)`.  Thoughts?

**Comment by eddelbuettel** [comment]
Looks like the second aspect can be fixed too similar to what I found [here](https://github.com/kaskr/RTMB/commit/5fe854d6a0ae75749010e6568fa562b18d84a57e) (gotta love github search with 'cran' ...)

**Comment by lsilvest** [comment]
This is the function called by `Rf_asS4`:

``` r
SEXP asS4(SEXP s, Rboolean flag, int complete)
{
    if(flag == IS_S4_OBJECT(s))
	return s;
    PROTECT(s);
    if(MAYBE_SHARED(s)) {
	s = shallow_duplicate(s);
	UNPROTECT(1);
	PROTECT(s);
    }
    if(flag) SET_S4_OBJECT(s);
    else {
	if(complete) {
	    SEXP value;
	    /* TENTATIVE:  how much does this change? */
	    if((value = R_getS4DataSlot(s, ANYSXP))
	       != R_NilValue && !IS_S4_OBJECT(value)) {
	      UNPROTECT(1);
	      return value;
	    }
	    /* else no plausible S3 object*/
	    else if(complete == 1) /* ordinary case (2, for conditional) */
	      error(_("object of class \"%s\" does not correspond to a valid S3 object"),
		      CHAR(STRING_ELT(R_data_class(s, FALSE), 0)));
	    else {
		UNPROTECT(1);
		return s; /*  unchanged */
	    }
	}
	UNSET_S4_OBJECT(s);
    }
    UNPROTECT(1);
    return s;
}
```

Things work for me when changing:
``` r
SET_S4_OBJECT(res);
```
to:
``` r 
res = Rf_asS4(res, TRUE, FALSE);
```
But not when there is no assignment to `res`. So it seems a shallow duplicate is made... I have to track down what `shallow_duplicate` does exactly to make sure we are not incurring a performance penalty when swapping to `Rf_asS4`.

**Comment by eddelbuettel** [comment]
That is ... in a way even more puzzling.  Maybe we can also step back and see if we can shimmy improvements into Rcpp::S4 which currently is fairly bare (but exists, mostly to get conveniently at a slot).

**Comment by eddelbuettel** [comment]
Yep. I got as far. Sadly the header info is hidden away.   The assignment trick is already pretty good.

**Comment by lsilvest** [comment]
So the reason `MAYBE_SHARED` returns true is that the ref count for the new object is indeed larger than 1. It doesn't really make sense because we've just allocated the vector. But it turns out the ref count is incremented to 3 after the copying of the names, so this code for finding the start of a `nanoival`:

``` c++
Rcpp::NumericVector nanoival_get_start_impl(const Rcpp::ComplexVector cv) {
  Rcpp::NumericVector res(cv.size());
  Rcpp::Rcout << "Ref count just after allocation: " << REFCNT(res) << std::endl;
  for (R_xlen_t i=0; i<cv.size(); ++i) {
    interval ival;
    Rcomplex c = cv[i];
    memcpy(&ival, reinterpret_cast<const char*>(&c), sizeof(c));
    if (ival.isNA()) {
      double d;
      memcpy(&d, reinterpret_cast<const char*>(&NA_INTEGER64), sizeof(NA_INTEGER64));
      res[i] = d;
    }
    else {
      std::int64_t start = ival.s;
      double d;
      memcpy(&d, reinterpret_cast<const char*>(&start), sizeof(start));
      res[i] = d;
    }
  }
  Rcpp::Rcout << "Ref count after filling in the start vector: " << REFCNT(res) << std::endl;
  res.names() = cv.names();
  Rcpp::Rcout << "Ref count after assigning the names: " << REFCNT(res) << std::endl;
  return assignS4("nanotime", res, "integer64");
}
```
when executed prints:
```
Ref count just after allocation: 1
Ref count after filling in the start vector: 1
Ref count after assigning the names: 3
```

I believe this is happening in `Rcc::NamesProxyPolicy`; in the `set` function, I see a few creations of `Shield<SEXP>` which would increase the ref count, but they should decrease when going out of scope. 

Anyway, a bit lost at this point. I will need to investigate closer the reference counting that is occurring in the names `set`. It is likely happening inside the R callbacks that occur in `NamesProxyPolicy::set`:

``` c++
void set(SEXP x) {
            Shield<SEXP> safe_x(x);

            /* check if we can use a fast version */
            if( TYPEOF(x) == STRSXP && parent.size() == Rf_length(x) ){
                Rf_namesgets(parent, x);
            } else {
                /* use the slower and more flexible version (callback to R) */
                SEXP namesSym = Rf_install( "names<-" );
                Shield<SEXP> call(Rf_lang3(namesSym, parent, x));
                Shield<SEXP> new_vec(Rcpp_fast_eval(call, R_GlobalEnv));
                parent.set__(new_vec);
            }

        }
```

Does any of this ring a bell?

I'm thinking we could call `assignS4` before copying the names therefore guaranteeing we don't do a shallow duplicate and the code will be equivalent. 

Thoughts?

**Comment by eddelbuettel** [comment]
I don't think Rcpp itself sets or increments the reference counters -- that is all deep down in R itself business by Luke.  Shield<> etc should only be nicer-to-use and simpler-because-RAII variants of PROTECT / UNPROTECT unless I miss something.  So moving assignS4 up an instruction or two seems fine to me.  Can you try it?

**Comment by lsilvest** [comment]
Tried it and it works. That's probably the best solution. Will go ahead and commit this, do you want me to also fix the missing \ in the `\code{+}` hoping Roxygen2 doesn't grumble on my machine? Or are you taking care of the latter fix?

**Comment by eddelbuettel** [comment]
Awesome.

I can take care of roxygen2, that has come up elsewhere and gotten fixed (incl. at work).

**Comment by eddelbuettel** [comment]
Confirming this local state of affairs passes muster:

```diff
modified   inst/include/nanotime/utilities.hpp
@@ -48,8 +48,8 @@ namespace nanotime {
   SEXP assignS4(const char* classname, Rcpp::Vector<R>& res) {
     Rcpp::CharacterVector cl = Rcpp::CharacterVector::create(classname);
     cl.attr("package") = "nanotime";
+    Rf_asS4(res, TRUE, FALSE);
     res.attr("class") = cl;
-    SET_S4_OBJECT(res);
     return Rcpp::S4(res);
   }
 
@@ -61,11 +61,37 @@ namespace nanotime {
     Rcpp::CharacterVector cl = Rcpp::CharacterVector::create(classname);
     cl.attr("package") = "nanotime";
     res.attr("class") = cl;
+    res = Rf_asS4(res, TRUE, FALSE);
     // use 'install' here as in R source code attrib.c: LLL
     Rcpp::CharacterVector oc = Rcpp::CharacterVector::create(oldClass);
     res.attr(".S3Class") = oc;
-    SET_S4_OBJECT(res);
     return Rcpp::S4(res);
```

**Comment by lsilvest** [comment]
The name copying also needs to be changed in `interval.cpp`. In `utilities.hpp` the order is unimportant. I've opened PR #121 for these changes. Maybe you can slip in there the `\code{+}` fix? I also see we've got #117 open. It's a small fix that I could slip in here unless you prefer to keep fixed separate?

**Comment by eddelbuettel** [comment]
No strong feelings.  Separate issues (as #121 and #117 seem to be) may warrant separate PRs...

I am getting quite used to squashing PRs (as we do at work, started it for Rcpp a while back too) which makes things nice and linear.  Order doesn't matter.  I can wrap up #121 and was about to suggest to ship to CRAN but we may to hit #117 over the head too.

**Comment by eddelbuettel** [comment]
BTW I checked the change under r-devel and I don't think i had the `interval.cpp` change in and now I confused why R didn't yell at me. Also I get no `\code{+}` hint either under (current, from this week) r-devel or r-release.

**Comment by lsilvest** [comment]
OK, I'll fix #117 tonight in a different PR and like that we can submit to CRAN only once. Without the change in `interval.cpp` things work, but then we get a shallow duplicate for no reason at all. Switching the order means no shallow duplicate because the ref count is still at 1 before the names assignment.

**Comment by eddelbuettel** [comment]
Ok, I will squash merge #121 then so you have clean slate to start from.  

Onwards, and forwards, and big big thanks for working through this one where I only managed half of it at best.

**Comment by eddelbuettel** [comment]
If found the '\code{+}' issue (reported at CRAN).  Might be easiest to remove the outer backticks around `\code{+}` in the Rd formating.

**Comment by eddelbuettel** [comment]
Naaa, I already fixed that issue weeks in [this commit](https://github.com/eddelbuettel/nanotime/commit/e5eac55d99defc7647948ae4bfc0388f7d519880). Sorry for a) not making that clearer and b) not going via a PR so that you saw it.
